chain/client/src/archive/cold_store_actor.rs

```diff
@@ -0,0 +1,496 @@
+                        tracing::info!(target : "cold_store", "Starting the cold store loop");
+                        self.cold_store_loop(ctx);
+                    }
+                }
+            }
+        }
+    }
+
+    // This method will copy data from hot storage to cold storage in a loop.
+    // It will try to copy blocks as fast as possible up until cold head = final head.
+    // Once the cold head reaches the final head it will sleep for one second before
+    // trying to copy data at the next height.
+    fn cold_store_loop(&self, ctx: &mut dyn DelayedActionRunner<Self>) {
+        if !self.keep_going.load(std::sync::atomic::Ordering::Relaxed) {
+            tracing::debug!(target : "cold_store", "Stopping the cold store loop");
+            return;
+        }
+
+        let result = self.cold_store_loop_impl();
+
+        // A block older than the final head was copied. We should continue copying
+        // until cold head reaches final head.
+        let dur = if let Ok(ColdStoreCopyResult::OtherBlockCopied) = result {
+            Duration::ZERO
+        } else {
+            self.split_storage_config.cold_store_loop_sleep_duration
+        };
+
+        ctx.run_later("cold_store_loop", dur, move |actor, ctx| {
+            actor.cold_store_loop(ctx);
+        });
+    }
+
+    fn cold_store_loop_impl(&self) -> anyhow::Result<ColdStoreCopyResult, ColdStoreError> {
+        let instant = std::time::Instant::now();
+        let result = self.cold_store_copy();
+        let duration = instant.elapsed();
+
+        let result_string = cold_store_copy_result_to_string(&result);
+        metrics::COLD_STORE_COPY_RESULT.with_label_values(&[result_string]).inc();
+        if duration > std::time::Duration::from_secs(1) {
+            tracing::debug!(target : "cold_store", "cold_store_copy took {}s", duration.as_secs_f64());
+        }
+
+        match &result {
+            Err(err) => {
+                tracing::error!(target : "cold_store", error = format!("{err:#?}"), "cold_store_copy failed");
+            }
+            Ok(copy_result) => match copy_result {
+                ColdStoreCopyResult::NoBlockCopied => {
+                    tracing::trace!(target: "cold_store", "No block was copied - cold head is up to date");
+                }
+                ColdStoreCopyResult::LatestBlockCopied => {
+                    tracing::trace!(target: "cold_store", "Latest block was copied");
+                }
+                ColdStoreCopyResult::OtherBlockCopied => {
+                    tracing::trace!(target: "cold_store", "Other block was copied - more copying needed");
+                }
+            },
+        }
+
+        result
+    }
+}
+
+/// Creates the cold store actor and keep_going handle if cold store is configured.
+/// Returns None if cold store is not configured or trie changes are not saved.
+/// This replaces the original `spawn_cold_store_loop` function.
+pub fn create_cold_store_actor(
```

Comment:
Is keep_going the idiomatic way to stop an actor? I thought actor would have it's own utilities for that. Not a biggie though.

Reply:
It's definitely not but I kept that to reduce the code change. As a follow up we can get rid of that and make things more like GC actor